apps/aecore/src/aec_miner.erl

```diff
@@ -133,128 +196,263 @@ init(_) ->
+    {keep_state, State#state{ miner = OtherMiner }};
+running(cast, create_block_candidate, State) ->
+    {next_state, configure, State, [postpone]};
+running(cast, bump_initial_cycle_nonce, State) ->
+    {next_state, configure, State, [postpone]};
 running(cast, mine, #state{block_candidate = BlockCandidate,
                            cycle_attempts_count = CycleAttemptsCount,
                            initial_cycle_nonce = CurrentMiningNonce,
-                           max_block_candidate_nonce = MaxMiningNonce} = State) ->
-    case aec_mining:mine(BlockCandidate, CycleAttemptsCount, CurrentMiningNonce, MaxMiningNonce) of
-        {ok, Block} ->
-            ws_handler:broadcast(miner, mined_block, [{height,
-                                                       aec_blocks:height(Block)}]),
-            ok = save_mined_block(Block),
-            gen_statem:cast(?SERVER, create_block_candidate),
-            {next_state, configure, State};
-        {error, generation_count_exhausted} ->
-            try exometer:update([ae,epoch,aecore,mining,retries], 1)
-            catch error:_ -> ok end,
-            epoch_mining:info("Failed to mine block in ~p attempts, retrying.",
-                              [CycleAttemptsCount]),
-            gen_statem:cast(?SERVER, bump_initial_cycle_nonce),
-            {next_state, configure, State};
-        {error, nonce_range_exhausted} ->
-            try exometer:update([ae,epoch,aecore,mining,retries], 1)
-            catch error:_ -> ok end,
-            epoch_mining:info("Failed to mine block, nonce range exhausted; regenerating block candidate"),
-            gen_statem:cast(?SERVER, create_block_candidate),
-            {next_state, configure, State}
-    end;
-running({call, From}, _Msg, State) ->
-    {next_state, running, State, [{reply, From, {error, not_supported}}]};
-running(cast, _Msg, State) ->
-    {next_state, idle, State}.
+                           max_block_candidate_nonce = MaxMiningNonce,
+                           miner = none}
+        = State) ->
+    Miner = mine(BlockCandidate, CycleAttemptsCount,
+                 CurrentMiningNonce, MaxMiningNonce,
+                 State),
+    {keep_state, State#state{ miner = Miner }};
+running(cast, mine, #state{miner = _Other}= State) ->
+    %% Already started mining
+    %% (Future implementations could allow this, starting from new nonce.)
+    epoch_mining:info("Trying to start while mining: ~p~n", [_Other]),
+    {keep_state, State, [postpone]};
+running(cast, check_keys, State) ->
+    %% The check_keys message should come when waiting.
+    epoch_mining:error("Got check_keys event in running state~n", []),
+    %% Keep the event and wait for a state transition.
+    {keep_state, State, [postpone]};
+%% Catchall, log & drop unwanted messages.
+running(_Type, _Msg, State) ->
+    epoch_mining:error("Unknown event in running: ~p:~p~n", [_Type,_Msg]),
+    {keep_state, running, State}.
```

Comment:
@gorillainduction Could you please clarify this return value? It appears to me you want `{keep_state, State}`.

Reply:
Indeed. This seems to be a bug. @happi can you comment?

Reply:
Although it is a bug, it is mostly left-over from cleaning up the messages in the gen_statem. This should never trigger.

I suggest we merge this, and I create a follow up ticket to fix this buglet to get the merge done today.